GHA: small improvements for Cygwin and Linux actions

MisterDA opened the pull request:
- [GHA: tell Cygwin's bash to ignore CR line endings](https://github.com/ocaml/ocaml/commit/a99f39569f1fc439ec883731f96f6286db24b8df) in workflow scripts.
  As [suggested by the cygwin-install action](https://github.com/cygwin/cygwin-install-action?tab=readme-ov-file#line-endings). Makes much nicer looking shell scripts! This also makes the build-msvc workflow look more like the 'regular' workflows, and Bash can report the line on which an error occurs, if any. Previously, the whole script was submitted as a single line.
- [GHA: use native Windows symlinks when installing Cygwin](https://github.com/ocaml/ocaml/commit/a583eafb7940b39ed1bcbb579768b0f003c5bf95): slightly more efficient install.
- [GHA: run all Make targets in parallel](https://github.com/ocaml/ocaml/commit/41c7e42aa2ec84fb4c130e22ad8d5ca3ada3ae76) for more efficient builds. Installation rules can also be parallelized. I've found it useful to limit the number of concurrent jobs with `nproc`. If unbounded, the out-of-memory (OOM) killer may be triggered on Linux.

(no change entry needed)

dra27 committed:
GHA: remove PRISTINE_PATH

Should be unnecessary

dra27 committed:
GHA: bash is part of Cygwin base

dra27 committed:
GHA: use a fold for config.log in GHA

If configure fails, it is useful to have config.log, but the error is
initially more useful, and that's often harder to spot at the end of
config.log

dra27 committed:
GHA: refactor configure calls in runner.sh

dra27 committed:
GHA: typo

MisterDA committed:
GHA: tell Cygwin's bash to ignore CR line endings

As suggested by the cygwin-install action. Makes much nicer looking
shell scripts!

https://github.com/cygwin/cygwin-install-action?tab=readme-ov-file#line-endings

MisterDA committed:
GHA: store the Autoconf cache key in the step output

Avoid using Cygwin when computing the cache key.

MisterDA committed:
GHA: don't set CC as an env variable

Makes a cleaner ./configure invocation.

dra27 commented on .github/workflows/build-msvc.yml:
```diff
@@ -74,11 +74,11 @@ jobs:
 
       - name: Compute a key to cache configure results
-        shell: bash
+        shell: bash -eo pipefail -o igncr {0}
```
I think this can instead be set as the default? (see [docs](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/setting-a-default-shell-and-working-directory)?

From a previous conversation with @shym, the `PRISTINE_PATH` thing isn't needed and could be removed - cf. https://github.com/dra27/ocaml/pull/173

MisterDA replied:
Nice. I also tried setting `SHELLOPTS`, but it [triggers problems in other scripts](https://github.com/cygwin/cygwin-install-action/pull/22). Maybe `BASHOPTS` would work instead. I'll use your solution now.

dra27 commented on .github/workflows/build-msvc.yml:
```diff
@@ -60,6 +60,8 @@ jobs:
       - name: Install Cygwin
         uses: cygwin/cygwin-install-action@v3
+        env:
+          CYGWIN: winsymlinks:nativestrict
```
Does this definitely work (have you checked the directory listing after installation?) I thought it remains necessary to pass `--symlink-type native` (cf. https://github.com/ocaml/opam/pull/5830). Is it significantly affecting the cache?

MisterDA replied:
The action documentation says:
> Alternatively, putting e.g. `CYGWIN=winsymlinks:native` into the workflow's environment works, since setup now honours that.

I haven't checked the cache or results yet. I've been setting this variable in my environment for a while with the only problem being that zsh refuses to write its history file with `nativestrict`.

dra27 replied:
Hmm, I'm not sure - I have memories in the (not too distant) past of symlinks causing strange issues elsewhere (while doing [dra27/cygsympathy](https://github.com/dra27/cygsympathy) IIRC). Given that this is just for the duration of the runner, and it's not necessarily the user's default configuration, I think it's better not to change this?

MisterDA replied:
Ok, I've removed this commit.

dra27 commented on .github/workflows/build-cross.yml:
```diff
@@ -52,8 +52,8 @@ jobs:
           if ! [ "$res" = 0 ]; then cat config.log; exit "$res"; fi
-          make -j
-          make install
+          make -j$(nproc)
```
Out of curiosity, have you _ever_ seen the OOM killer terminate a single build of OCaml on anything? When dependencies are being compiled, it's definitely beneficial to be running more jobs than a simple CPU count would imply, cf. this little bit of history in the AppVeyor script from #10049 (note that it was necessary _on a two vCPU Windows machine_ to limit the main build, but the `-j64` was added for the deps run because otherwise it became intolerably slow):
https://github.com/ocaml/ocaml/blob/ebb54c629a17d9041eba55b7c3a0089f0700d02b/tools/ci/appveyor/appveyor_build.sh#L216-L219

MisterDA replied:
> Out of curiosity, have you ever seen the OOM killer terminate a single build of OCaml on anything?

I have! In the cross-compiler job using mingw-w64 on Linux. The GCC processes are terminated. I think it's the OOM.
https://github.com/MisterDA/ocaml/actions/runs/13941725261/job/39019956117
```
x86_64-w64-mingw32-gcc: fatal error: Killed signal terminated program cc1
compilation terminated.
make[2]: *** [../Makefile.otherlibs.common:164: truncate_win32.b.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [../Makefile.otherlibs.common:164: access.b.o] Terminated
make[2]: *** [../Makefile.otherlibs.common:164: addrofstr.b.o] Terminated
...
```

I can try setting the limit to _n_ times _nproc_, but setting an unbounded number of jobs _definitely_ triggers something that terminates processes on this job.

dra27 replied:
I stand corrected! Please could you add a comment to that effect?

MisterDA replied:
I have added:

> The OOM-killer may be triggered if the number of parallel jobs isn't limited.

dra27 commented on .github/workflows/build-cross.yml:
```diff
@@ -90,8 +90,8 @@ jobs:
-          make crossopt -j$(nproc)
-          make installcross
+          make -j$(nproc) crossopt
+          make -j$(nproc) installcross
```
The `install` targets don't have any dependencies - what are they supposed to be doing in parallel?

MisterDA replied:
Right.

dra27 commented on .github/workflows/build-msvc.yml:
```diff
@@ -136,8 +136,8 @@ jobs:
         run: |
           eval $(tools/msvs-promote-path)
-          make -j tests
+          make -j$(nproc) tests
```
These days, this shouldn't be doing anything anyway (the testsuite components and libraries are built as part of the main build since #12706) - perhaps just delete the `-j` instead?

MisterDA replied:
Right.

dra27 reviewed (commented):
I like the workflow tidying up parts (especially making the scripts clearer) - I'm not so convinced by the symlinks/parallelism bits.

I did some messing around in November in dra27/ocaml#173, which you're welcome to cherry-pick/adapt any or none of!

MisterDA committed:
GHA: split configuration and build

Matches the other workflows.

dra27 commented on .github/workflows/build-msvc.yml:
```diff
@@ -101,7 +101,7 @@ jobs:
             if ((failed)) ; then cat config.log ; exit $failed ; fi
           fi
-          make -j || failed=$?
+          make -j$(nproc) || failed=$?
```
Assuming this one hasn't failed before, though, can this one be left as-is (my concern is that we unexpectedly make this slower, without actually fixing anything - I'm less worried in the build-cross workflow about it being slower, because that workflow is only executed on-demand)

MisterDA replied:
Left as-is.

MisterDA committed:
GHA: revisit limits on concurrent Make jobs

- When using MinGW-w64 on Linux, Make may spawn too many processes,
  triggering the OOM-killer, and ending the build.
- Running `make test` with `-j` has no effect, the test runner is
  already built and is responsible of running tests in parallel.

dra27 reviewed (approved):
LGTM, thanks!
